sys/usb/usbus/cdc/ecm/cdc_ecm.c

```diff
@@ -0,0 +1,386 @@
+    usb_desc_ecm_t ecm;
+
+    /* functional cdc ecm descriptor */
+    ecm.length = sizeof(usb_desc_ecm_t);
+    ecm.type = USB_TYPE_DESCRIPTOR_CDC;
+    ecm.subtype = USB_CDC_DESCR_SUBTYPE_ETH_NET;
+    ecm.macaddress = cdcecm->mac_str.idx;
+    ecm.ethernetstatistics = 0;
+    ecm.maxsegmentsize = ETHERNET_FRAME_LEN;
+    ecm.numbermcfilters = 0x0000; /* No filtering */
+    ecm.numberpowerfilters = 0;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t *)&ecm, sizeof(ecm));
+    return sizeof(usb_desc_ecm_t);
+}
+
+static size_t _gen_cdc_descriptor(usbus_t *usbus, void *arg)
+{
+    (void)arg;
+    usb_desc_cdc_t cdc;
+    /* functional cdc descriptor */
+    cdc.length = sizeof(usb_desc_cdc_t);
+    cdc.bcd_hid = USB_CDC_VERSION_BCD;
+    cdc.type = USB_TYPE_DESCRIPTOR_CDC;
+    cdc.subtype = 0x00;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t *)&cdc, sizeof(cdc));
+    return sizeof(usb_desc_cdc_t);
+}
+
+static void _notify_link_speed(usbus_cdcecm_device_t *cdcecm)
+{
+    DEBUG("CDC ECM: sending link speed indication\n");
+    usb_setup_t *notification = (usb_setup_t *)cdcecm->ep_ctrl->ep->buf;
+    notification->type = USB_SETUP_REQUEST_DEVICE2HOST |
+                         USB_SETUP_REQUEST_TYPE_CLASS |
+                         USB_SETUP_REQUEST_RECIPIENT_INTERFACE;
+    notification->request = USB_CDC_MGNT_NOTIF_CONN_SPEED_CHANGE;
+    notification->value = 0;
+    notification->index = cdcecm->iface_ctrl.idx;
+    notification->length = 8;
+
+    uint32_t *speed =
+        (uint32_t *)(cdcecm->ep_ctrl->ep->buf + sizeof(usb_setup_t));
+    *speed++ = USBUS_CDC_ECM_CONFIG_SPEED_DOWNSTREAM;
+    *speed = USBUS_CDC_ECM_CONFIG_SPEED_UPSTREAM;
+    usbdev_ep_ready(cdcecm->ep_ctrl->ep,
+                    sizeof(usb_setup_t) + 2 * sizeof(uint32_t));
+    cdcecm->notif = USBUS_CDCECM_NOTIF_SPEED;
+}
+
+static void _notify_link_up(usbus_cdcecm_device_t *cdcecm)
+{
+    DEBUG("CDC ECM: sending link up indication\n");
+    usb_setup_t *notification = (usb_setup_t *)cdcecm->ep_ctrl->ep->buf;
+    notification->type = USB_SETUP_REQUEST_DEVICE2HOST |
+                         USB_SETUP_REQUEST_TYPE_CLASS |
+                         USB_SETUP_REQUEST_RECIPIENT_INTERFACE;
+    notification->request = USB_CDC_MGNT_NOTIF_NETWORK_CONNECTION;
+    notification->value = 1;
+    notification->index = cdcecm->iface_ctrl.idx;
+    notification->length = 0;
+    usbdev_ep_ready(cdcecm->ep_ctrl->ep, sizeof(usb_setup_t));
+    cdcecm->notif = USBUS_CDCECM_NOTIF_LINK_UP;
+}
+
+static const usbus_handler_driver_t cdcecm_driver = {
+    .init = _init,
+    .event_handler = _event_handler,
+    .transfer_handler = _transfer_handler,
+    .setup_handler = _setup_handler,
+};
+
+static void _fill_ethernet(usbus_cdcecm_device_t *cdcecm)
+{
+    uint8_t ethernet[ETHERNET_ADDR_LEN];
+
+    luid_get(ethernet, ETHERNET_ADDR_LEN);
+    ethernet[0] |= 0x02;        /* locally administered address */
```

Comment:
blackmagic ?

Reply:
Taken from [here](https://github.com/RIOT-OS/RIOT/blob/master/drivers/enc28j60/enc28j60.c#L430), forces the ethernet mac locally administrated bit on as the MAC address is generated by the LUID module.

Reply:
[source](http://www.ieee802.org/secmail/pdfocSP2xXA6d.pdf) (section 9.2, Figure 8)

Reply:
see #11666

Reply:
Changed to use the functions added in #11666